File: OJ/usaco/s2/s24/ttwo.pas
{
ID:lijianx1
PROG:ttwo
LANG:PASCAL
}
program ttwo;
const maxn=10;
      dx:array[1..4] of longint=(-1,0,1,0);
      dy:array[1..4] of longint=(0,1,0,-1);
type node=record
         x,y,k:longint;
     end;
var ans:longint;
    flag:array[1..maxn,1..maxn] of boolean;
    a:array[1..2] of node;
procedure init;
 var i,j:longint;
     ch:char;
 begin
     assign(input,'ttwo.in'); reset(input);
     fillchar(flag,sizeof(flag),true);
     for i:=1 to maxn do begin
         for j:=1 to maxn do begin
             read(ch);
             case ch of
                 'C':begin
                         a[1].x:=i; a[1].y:=j; a[1].k:=1;
                     end;
                 'F':begin
                         a[2].x:=i; a[2].y:=j; a[2].k:=1;
                     end;
                 '*':flag[i,j]:=false;
             end;
         end;
         readln;
     end;
     close(input);
 end;
 
procedure main;
 var i,x,y:longint;
 begin
     ans:=0;
     repeat
        for i:=1 to 2 do begin
            x:=a[i].x+dx[a[i].k];
            y:=a[i].y+dy[a[i].k];
            if (x>=1) and (x<=maxn) and (y>=1) and (y<=maxn) and (flag[x,y]) then begin
                a[i].x:=x; a[i].y:=y;
            end else a[i].k:=a[i].k mod 4+1;
        end;
        inc(ans);
     until ((a[1].x=a[2].x) and (a[1].y=a[2].y)) or (ans>160000);
     assign(output,'ttwo.out'); rewrite(output);
     if (a[1].x=a[2].x) and (a[1].y=a[2].y) then writeln(ans) else writeln(0);
     close(output);
 end;
 
begin
    init;
    main;
end.

File: OJ/usaco/cpp/ttwo.cpp
/*
ID:oifox20071
LANG:C++
PROG:ttwo
*/
#include <iostream>
#include <cstring>
using namespace std;
const int n = 10;
const int maxans = 160000;
const int dx[4] = {-1, 0, 1, 0};
const int dy[4] = {0, 1, 0, -1};
struct node
{
	int x, y, direction;
} farmer, cow;
bool flag[n][n];
void init()
{
	freopen("ttwo.in", "r", stdin); freopen("ttwo.out", "w", stdout);
	char ch;
	memset(flag, true, sizeof(flag));
	for (int i = 0; i < n; ++ i)
		for (int j = 0; j < n; ++ j)
		{
			cin >> ch;
			switch (ch)
			{
				case 'C':
					cow.x = i; cow.y = j; cow.direction = 0; break;
				case 'F':
					farmer.x = i; farmer.y = j; farmer.direction = 0; break;
				case '*':
					flag[i][j] = false; break;
			}
		}
	
}
int main()
{
	init();
	int ans = 0;
	do
	{
		int x = cow.x + dx[cow.direction];
		int y = cow.y + dy[cow.direction];
		if (x >= 0 && x < n && y >= 0 && y < n && flag[x][y])
		{
			cow.x = x; cow.y = y;
		}
		else cow.direction = (cow.direction + 1) % 4;
		
		x = farmer.x + dx[farmer.direction];
		y = farmer.y + dy[farmer.direction];
		if (x >= 0 && x < n && y >= 0 && y < n && flag[x][y])
		{
			farmer.x = x; farmer.y = y;
		}
		else farmer.direction = (farmer.direction + 1) % 4;
		++ ans;
	} while ((cow.x != farmer.x || cow.y != farmer.y) && ans <= maxans);
	if (cow.x == farmer.x && cow.y == farmer.y)
		cout << ans << endl;
	else
		cout << 0 << endl;
	return 0;
}
